specification/sections/04-hooks.md

```diff
@@ -81,75 +87,154 @@ see: [dynamic-context paradigm](../glossary.md#dynamic-context-paradigm)
 
 > Condition: The provider `metadata` field in the `hook context` **MUST** be immutable.
 
-### 4.3. Hook creation and parameters
+### 4.3. Hook data
+
+Hook data exists to allow hook stages to share data for a specific evaluation. For instance a span
+for OpenTelemetry could be created in a `before` stage and closed in an `after` stage.
+
+```Java
+  public Optional<EvaluationContext> before(HookContext context, HookHints hints, HookData data) {
+    SpanBuilder builder = tracer.spanBuilder('sample')
+    .setParent(Context.current().with(Span.current()));
+    Span span = builder.startSpan()
+    data.set("span", span);
+  }
+
+  public void after(HookContext context, FlagEvaluationDetails details, HookHints hints, HookData data) {
+    // Only accessible by this hook for this specific evaluation.
+    Object value = data.get("span");
+    if (value instanceof Span) {
+      Span span = (Span) value;
+      span.end();
+    }
+  }
+```
 
 #### Requirement 4.3.1
 
+> `hook data` **MUST** be a structure supports definition of arbitrary properties, with keys of type `string`, and values of type `boolean | string | number | datetime | structure`.
+
+### 4.4. Hook creation and parameters
+
+#### Requirement 4.4.1
+
 > Hooks **MUST** specify at least one stage.
 
-#### Condition 4.3.2
+#### Requirement 4.4.2
+
+> `Hook data` **MUST** must be created before the first `stage` invoked in a hook for a specific evaluation and propagated between each `stage` of the hook.
```

Comment:
I think it's implied here, but not explicit, that hook data is scoped only to the hook instance in question. I think it might be worth adding a specific point about that, and perhaps even a recommendation or example of how it might be accomplished in some implementations.

Reply:
I added some more clarifying language a couple places.